File: src/bitboard.rs
#[derive(Debug, Clone, Copy)]
pub struct Bitboard {
    pub w_pawn: u64,
    pub w_knight: u64,
    pub w_bishop: u64,
    pub w_rook: u64,
    pub w_queen: u64,
    pub w_king: u64,
    pub b_pawn: u64,
    pub b_knight: u64,
    pub b_bishop: u64,
    pub b_rook: u64,
    pub b_queen: u64,
    pub b_king: u64,
}

impl Bitboard {
    pub fn new() -> Self {
        Bitboard {
            w_pawn: 0,
            w_knight: 0,
            w_bishop: 0,
            w_rook: 0,
            w_queen: 0,
            w_king: 0,
            b_pawn: 0,
            b_knight: 0,
            b_bishop: 0,
            b_rook: 0,
            b_queen: 0,
            b_king: 0,
        }
    }

    pub fn init(&mut self) {
        self.w_pawn = 0xFF << 48;
        self.w_knight = 0x42 << 56;
        self.w_bishop = 0x24 << 56;
        self.w_rook = 0x81 << 56;
        self.w_queen = 0x08 << 56;
        self.w_king = 0x10 << 56;

        self.b_pawn = 0xFF << 8;
        self.b_knight = 0x42;
        self.b_bishop = 0x24;
        self.b_rook = 0x81;
        self.b_queen = 0x08;
        self.b_king = 0x10;
    }

    pub fn clear(&mut self) {
        *self = Bitboard::new();
    }

    pub fn get_occupied_squares(&self) -> u64 {
        self.w_pawn | self.w_knight | self.w_bishop | self.w_rook | self.w_queen | self.w_king |
        self.b_pawn | self.b_knight | self.b_bishop | self.b_rook | self.b_queen | self.b_king
    }

    pub fn get_empty_squares(&self) -> u64 {
        !self.get_occupied_squares()
    }

    pub fn get_white_squares(&self) -> u64 {
        self.w_pawn | self.w_knight | self.w_bishop | self.w_rook | self.w_queen | self.w_king
    }

    pub fn get_black_squares(&self) -> u64 {
        self.b_pawn | self.b_knight | self.b_bishop | self.b_rook | self.b_queen | self.b_king
    }

    pub fn to_string(&self) -> String {
        let mut board = vec!['-'; 72];
        for i in 0..8 {
            board[i * 9 + 8] = '\n';
        }

        let piece_names = ['P', 'N', 'B', 'R', 'Q', 'K', 'p', 'n', 'b', 'r', 'q', 'k'];
        let pieces = [
            self.w_pawn, self.w_knight, self.w_bishop, self.w_rook, self.w_queen, self.w_king,
            self.b_pawn, self.b_knight, self.b_bishop, self.b_rook, self.b_queen, self.b_king,
        ];

        for (piece, &bitboard) in piece_names.iter().zip(pieces.iter()) {
            for i in 0..64 {
                if (bitboard >> i) & 1 == 1 {
                    let rank = 7 - (i / 8);
                    let file = i % 8;
                    let pos = rank * 9 + file;
                    board[pos] = *piece;
                }
            }
        }

        board.iter().collect()
    }

    pub fn from_fen(fen: &str) -> Self {
        let mut board = Bitboard::new();
        let mut parts = fen.split_whitespace();
        let position = parts.next().unwrap();
        let ranks: Vec<&str> = position.split('/').collect();
        
        for (rank_idx, rank) in ranks.iter().enumerate() {
            let mut file = 0;
            for c in rank.chars() {
                if c.is_digit(10) {
                    file += c.to_digit(10).unwrap() as usize;
                } else {
                    let square = (rank_idx * 8 + file) as u64;
                    match c {
                        'P' => board.w_pawn |= 1 << square,
                        'N' => board.w_knight |= 1 << square,
                        'B' => board.w_bishop |= 1 << square,
                        'R' => board.w_rook |= 1 << square,
                        'Q' => board.w_queen |= 1 << square,
                        'K' => board.w_king |= 1 << square,
                        'p' => board.b_pawn |= 1 << square,
                        'n' => board.b_knight |= 1 << square,
                        'b' => board.b_bishop |= 1 << square,
                        'r' => board.b_rook |= 1 << square,
                        'q' => board.b_queen |= 1 << square,
                        'k' => board.b_king |= 1 << square,
                        _ => {}
                    }
                    file += 1;
                }
            }
        }
        board
    }
} 

File: go_engine/go_engine/chess/bitboard.go
package chess

// Bitboard represents a chess position using bitboards for each piece type
type Bitboard struct {
	Wpawn   uint64
	Wknight uint64
	Wbishop uint64
	Wrook   uint64
	Wqueen  uint64
	Wking   uint64

	Bpawn   uint64
	Bknight uint64
	Bbishop uint64
	Brook   uint64
	Bqueen  uint64
	Bking   uint64
}

// PieceNames maps piece types to their FEN notation characters
var PieceNames = []rune{'p', 'n', 'b', 'r', 'q', 'k', 'P', 'N', 'B', 'R', 'Q', 'K'}

// InitBitboard initializes a new bitboard with pieces in their starting positions
func InitBitboard() *Bitboard {
	bb := &Bitboard{}
	bb.Wpawn = uint64(255) << 48  // White pawns on rank 2
	bb.Wknight = uint64(66) << 56 // White knights on b1 and g1
	bb.Wbishop = uint64(36) << 56 // White bishops on c1 and f1
	bb.Wrook = uint64(129) << 56  // White rooks on a1 and h1
	bb.Wqueen = uint64(8) << 56   // White queen on d1
	bb.Wking = uint64(16) << 56   // White king on e1

	bb.Bpawn = uint64(255) << 8 // Black pawns on rank 7
	bb.Bknight = uint64(66)     // Black knights on b8 and g8
	bb.Bbishop = uint64(36)     // Black bishops on c8 and f8
	bb.Brook = uint64(129)      // Black rooks on a8 and h8
	bb.Bqueen = uint64(8)       // Black queen on d8
	bb.Bking = uint64(16)       // Black king on e8

	return bb
}

// ClearBitboard sets all bitboards to 0
func (bb *Bitboard) ClearBitboard() {
	bb.Wpawn = 0
	bb.Wknight = 0
	bb.Wbishop = 0
	bb.Wrook = 0
	bb.Wqueen = 0
	bb.Wking = 0

	bb.Bpawn = 0
	bb.Bknight = 0
	bb.Bbishop = 0
	bb.Brook = 0
	bb.Bqueen = 0
	bb.Bking = 0
}

// GetOccupiedSquares returns a bitboard with all occupied squares
func (bb *Bitboard) GetOccupiedSquares() uint64 {
	return bb.Wpawn | bb.Wknight | bb.Wbishop | bb.Wrook | bb.Wqueen | bb.Wking |
		bb.Bpawn | bb.Bknight | bb.Bbishop | bb.Brook | bb.Bqueen | bb.Bking
}

// GetEmptySquares returns a bitboard with all empty squares
func (bb *Bitboard) GetEmptySquares() uint64 {
	return ^bb.GetOccupiedSquares()
}

// GetWhiteSquares returns a bitboard with all squares occupied by white pieces
func (bb *Bitboard) GetWhiteSquares() uint64 {
	return bb.Wpawn | bb.Wknight | bb.Wbishop | bb.Wrook | bb.Wqueen | bb.Wking
}

// GetBlackSquares returns a bitboard with all squares occupied by black pieces
func (bb *Bitboard) GetBlackSquares() uint64 {
	return bb.Bpawn | bb.Bknight | bb.Bbishop | bb.Brook | bb.Bqueen | bb.Bking
}

// String returns a string representation of the board
func (bb *Bitboard) String() string {
	board := make([]rune, 71)
	for i := range board {
		if (i+1)%9 == 0 {
			board[i] = '\n'
		} else {
			board[i] = '-'
		}
	}

	for square := 0; square < 64; square++ {
		var piece rune
		switch {
		case (bb.Wpawn>>square)&1 == 1:
			piece = 'p'
		case (bb.Wknight>>square)&1 == 1:
			piece = 'n'
		case (bb.Wbishop>>square)&1 == 1:
			piece = 'b'
		case (bb.Wrook>>square)&1 == 1:
			piece = 'r'
		case (bb.Wqueen>>square)&1 == 1:
			piece = 'q'
		case (bb.Wking>>square)&1 == 1:
			piece = 'k'
		case (bb.Bpawn>>square)&1 == 1:
			piece = 'P'
		case (bb.Bknight>>square)&1 == 1:
			piece = 'N'
		case (bb.Bbishop>>square)&1 == 1:
			piece = 'B'
		case (bb.Brook>>square)&1 == 1:
			piece = 'R'
		case (bb.Bqueen>>square)&1 == 1:
			piece = 'Q'
		case (bb.Bking>>square)&1 == 1:
			piece = 'K'
		default:
			continue
		}
		board[square+square/8] = piece
	}

	return string(board)
}
